// vcloud-vms-service/lib/onvif/index.js
const onvif = require("node-onvif");

const { COMMANDS } = require("./config");

class ONVIFCamera {
  constructor() {
    this.camera = null;
  }

  async makeStreamURL({ ip, port, protocol }) {
    try {
      let xaddr = `${protocol}://${ip}:${port || 80}/onvif/device_service`;

      if (!port) {
        const foundCameras = await onvif.startProbe();
        console.log("LOCAL CAMERAS:", foundCameras);

        if (!foundCameras.length) {
          throw new Error("Can't find any camera !");
        }

        for (let i = 0; i < foundCameras.length; i++) {
          const isExistXaddr = foundCameras[i].xaddrs.find((xaddr) =>
            xaddr.includes(ip)
          );
          if (isExistXaddr) {
            xaddr = isExistXaddr;
            break;
          }
        }
      }

      return xaddr;
    } catch (err) {
      console.log(`MADE XADDR FAILED ${err.message}`);
      throw new Error(err.message);
    }
  }

  async init(options) {
    try {
      console.log(
        `INIT ONVIF START: PROTOCOL = ${options.protocol} IP = ${options.ip} PORT = ${options.port}`
      );

      const xaddr = await this.makeStreamURL(options);
      console.log(`XADDR: ${xaddr}`);
      this.camera = new onvif.OnvifDevice({
        xaddr,
        user: options.user,
        pass: options.pass,
        port: options.port,
        protocol: `${options.protocol}:`,
      });
      await this.camera.init();
      console.log("INIT ONVIF COMPLETE");
      // let streamUrl = this.camera.getUdpStreamUrl();
    } catch (err) {
      console.log(`INIT ONVIF FAILED ${err.message}`);
      // throw new Error(err.message);
    }
  }

  async move(command) {
    // Move the camera
      this.camera
          .ptzMove({ speed: COMMANDS[command], timeout: 1 })
          .then()
          .catch(err => console.log(err.message));

    if (command.includes('zoom')) {
      await new Promise(res => setTimeout(res, 250));
      this.stop();
    }
  }

  stop() {
    // Stop the camera
    this.camera.ptzStop().then().catch(err => console.log(err.message));
  }
}

module.exports = ONVIFCamera;
